### ISSUE TITLE
Dedicated AC Power Flow Solver

### ISSUE THREAD

**Issue by jchen975** [lanl-ansi/PowerModels.jl#590]
Hi, I just realized this problem, but when I run a solved Matpower case, it takes the same number of iterations as it would with default values. In Matpower, 
```
results = runpf('caseX') 
savecase('caseX', results)
runpf('caseX')
```
should result in `iterations: 1` in the end. When running with default/solved values for case118, and case300, it takes 4 and 5 iterations respectively. I tried both with the case saved from Matpower and with manually updating the `vm`, `va` fields of all buses from the solutions dict, and got the same results.

Any idea why this is happening? I know Matpower uses NR for acpf by default, so I'm assuming it has something to do with Ipopt. I'd appreciate any clarifications on this.

**Comment by ccoffrin** [comment]
This is a good question and there are quite a few details to get the most out of a warm start in PowerModels.  At a high level these are the key points,
1. you have to provide special values in the data model for providing an initial starting point.  In the case of the ACPPowerModel these are `pg_start`, `qg_start`, `vm_start`, `va_start`, `pf_start`, `pt_start`, `qf_start`, `qt_start` (note that the ACPPowerModel adds explicit flow variables to the branches).  `<var name>_start` is the pattern for all variables.
2. After setting these values you then need to set Ipopt's parameters carefully.  This includes setting smaller values for `mu_init`, `bound_frac` and `bound_push`.  For very large models you might also try `hessian_approximation` to reduce the time for derivative computations.

I hope that helps.  I'll also note that PowerModels was originally designed for solving OPF without a good guess for a starting point.  Due to that focus, doing very fast PF solves is not super easy in PowerModels.  In the future I plan to add some dedicated AC-PF solvers that assume a good starting point and are much faster and simpler to use for that use case.

**Comment by jchen975** [comment]
Hi Dr. Coffrin, regarding your previous comment on your plan to add an ACPF solver that accommodates warm-start conditions, has there been any updates? If that's not a short-term priority, I was actually thinking about creating a new package that only solves the PF problem, by rewriting the Matpower `runpf` related functions in Julia, sometime early next year (I'm finishing my undergrad this coming April so I probably won't be able to finish it while I'm still in school). Unfortunately I eventually switched back to Matpower to continue with my project, but I'd like to get your opinions on this. Thanks!

**Comment by ccoffrin** [comment]
@jchen975 there are not updates per-se, it is still possible to warm-start Ipopt with the current version of PowerModels, although it is not well documented.

Please feel free you write your own power flow solver, I hope to implement something similar in PowerModels in the next 6-12 months, so I might draw some inspiration from your implementation.

**Comment by ccoffrin** [comment]
a native DC power flow solver is not available.  The `AdmittanceMatrix` data structure as been added to help support this feature.

**Comment by timueh** [comment]
Hi @jchen975, is this still an open issue? I'd be interested in a dedicated power flow solver in Julia that uses `PowerModels.jl`. A first step might be to set up the system of nonlinear equations, and then use it with one of Julia's Newton-type solvers. Then, one could go about implementing decoupled power flow etc.. 

Any thoughts?

**Comment by ccoffrin** [comment]
FYI, @jd-lara as I know you have an interest in this.

**Comment by jchen975** [comment]
@timueh Yeah, unfortunately I haven't had the time to get on this... I was going to start in March once I finish a few other things on my hand. But what you suggested made sense! 

@ccoffrin Please feel free to close this issue :)

**Comment by timueh** [comment]
Sweet! Perhaps there is room for collaboration @jchen975? Perhaps we can talk about details via e-mail? You can find mine on my github profile.

**Comment by jd-lara** [comment]
@timueh we have implemented something similar in PowerSystems.jl where we use Julia's solvers to solve the Non-linear system of equations. The intention for us is to initialize the system with a PowerFlow solution when needed. 

https://github.com/NREL/PowerSystems.jl/tree/master/src/utils/power_flow

The tests show examples of how to use it. 

https://github.com/NREL/PowerSystems.jl/blob/master/test/test_powerflow.jl

**Comment by timueh** [comment]
Thanks. That looks like what I was looking for. Just a quick question---which I could answer myself with playing with the code, but you might know the answer immediately---what's the number of decision variables in the power flow problem? Is it `n_{pv} + 2*n_{pq}` like in matpower, where `n_{pv}` is the number of PV buses, and `n_{pq}` is the number of PQ. buses?

**Comment by jd-lara** [comment]
The number of decision variables is alway `2*n_buses`. Each PV bus needs a solution for \theta and Q. Each PQ bus needs a solution for \theta and V. Each Slack bus needs a solution for P and Q.

**Comment by ccoffrin** [comment]
A prototype version of this feature is now available in the `rc-v0.16` branch.  The function is called `compute_ac_pf`.  Feedback is welcome.

**Comment by ccoffrin** [comment]
After a long wait, I am happy to report that this has been implemented.  Documentation is available here, https://lanl-ansi.github.io/PowerModels.jl/stable/power-flow/
